File: ainpc-core-plugin/src/test/java/ro/ainpc/engine/runtime/ScenarioRuntimeRegistryTest.java
package ro.ainpc.engine.runtime;

import org.junit.jupiter.api.Test;

import java.util.LinkedHashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ScenarioRuntimeRegistryTest {

    @Test
    void actionRegistryNormalizesLookupAndRejectsDuplicateHandlers() {
        ScenarioActionRegistry registry = new ScenarioActionRegistry();

        ScenarioValidationReport first = registry.register(actionHandler("Give_Item"));
        ScenarioValidationReport duplicate = registry.register(actionHandler(" give_item "));

        assertTrue(first.isValid());
        assertFalse(duplicate.isValid());
        assertTrue(registry.supports(" GIVE_ITEM "));
        assertEquals(1, registry.handlers().size());
    }

    @Test
    void validateActionReportsMissingHandler() {
        ScenarioActionRegistry registry = new ScenarioActionRegistry();
        ScenarioRuntimeDefinition action = new ScenarioRuntimeDefinition(
            "reward-1",
            "give_item",
            Map.of("material", "EMERALD")
        );

        ScenarioValidationReport report = registry.validateAction(action);

        assertFalse(report.isValid());
        assertTrue(report.errors().getFirst().contains("give_item"));
    }

    @Test
    void validateActionAcceptsRegisteredHandlerAndWarnsForMissingStableId() {
        ScenarioActionRegistry registry = new ScenarioActionRegistry();
        registry.register(actionHandler("give_item"));
        ScenarioRuntimeDefinition action = new ScenarioRuntimeDefinition(
            "",
            "GIVE_ITEM",
            Map.of("material", "EMERALD")
        );

        ScenarioValidationReport report = registry.validateAction(action);

        assertTrue(report.isValid());
        assertTrue(report.hasWarnings());
    }

    @Test
    void executionContextAndDefinitionSanitizeMaps() {
        Map<String, String> variables = new LinkedHashMap<>();
        variables.put(" region ", " spawn ");
        variables.put("", "ignored");
        variables.put(null, "ignored");
        variables.put("missing", null);

        ScenarioExecutionContext context = new ScenarioExecutionContext(
            " player-uuid ",
            " Hero ",
            " npc-1 ",
            " Guard ",
            " region-1 ",
            " place-1 ",
            " node-1 ",
            " Q01 ",
            " progress-1 ",
            " standalone ",
            variables
        );

        ScenarioRuntimeDefinition definition = new ScenarioRuntimeDefinition(
            " action-1 ",
            " give_item ",
            variables
        );

        assertEquals("player-uuid", context.playerUuid());
        assertEquals("spawn", context.variable("region"));
        assertEquals("", context.variable("missing"));
        assertEquals("action-1", definition.id());
        assertEquals("give_item", definition.type());
        assertEquals("spawn", definition.parameter("region"));
        assertThrows(UnsupportedOperationException.class, () -> context.variables().put("new", "value"));
        assertThrows(UnsupportedOperationException.class, () -> definition.parameters().put("new", "value"));
    }

    private ScenarioActionHandler actionHandler(String type) {
        return new ScenarioActionHandler() {
            @Override
            public String type() {
                return type;
            }

            @Override
            public void execute(ScenarioExecutionContext context, ScenarioRuntimeDefinition action) {
            }
        };
    }
}

File: ainpc-core-plugin/bin/test/ro/ainpc/engine/runtime/ScenarioRuntimeRegistryTest.kt
package ro.ainpc.engine.runtime

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.util.LinkedHashMap

class ScenarioRuntimeRegistryTest {
    @Test
    fun actionRegistryNormalizesLookupAndRejectsDuplicateHandlers() {
        val registry = ScenarioActionRegistry()

        val first = registry.register(actionHandler("Give_Item"))
        val duplicate = registry.register(actionHandler(" give_item "))

        assertTrue(first.isValid())
        assertFalse(duplicate.isValid())
        assertTrue(registry.supports(" GIVE_ITEM "))
        assertEquals(1, registry.handlers().size)
    }

    @Test
    fun validateActionReportsMissingHandler() {
        val registry = ScenarioActionRegistry()
        val action = ScenarioRuntimeDefinition("reward-1", "give_item", mapOf("material" to "EMERALD"))

        val report = registry.validateAction(action)

        assertFalse(report.isValid())
        assertTrue(report.errors().first().contains("give_item"))
    }

    @Test
    fun validateActionAcceptsRegisteredHandlerAndWarnsForMissingStableId() {
        val registry = ScenarioActionRegistry()
        registry.register(actionHandler("give_item"))
        val action = ScenarioRuntimeDefinition("", "GIVE_ITEM", mapOf("material" to "EMERALD"))

        val report = registry.validateAction(action)

        assertTrue(report.isValid())
        assertTrue(report.hasWarnings())
    }

    @Test
    fun executionContextAndDefinitionSanitizeMaps() {
        val variables = LinkedHashMap<String, String>()
        variables[" region "] = " spawn "
        variables[""] = "ignored"
        variables["missing"] = ""

        val context = ScenarioExecutionContext(
            " player-uuid ", " Hero ", " npc-1 ", " Guard ", " region-1 ",
            " place-1 ", " node-1 ", " Q01 ", " progress-1 ", " standalone ", variables
        )

        val definition = ScenarioRuntimeDefinition(" action-1 ", " give_item ", variables)

        assertEquals("player-uuid", context.playerUuid())
        assertEquals("spawn", context.variable("region"))
        assertEquals("", context.variable("missing"))
        assertEquals("action-1", definition.id())
        assertEquals("give_item", definition.type())
        assertEquals("spawn", definition.parameter("region"))
        assertThrows(UnsupportedOperationException::class.java) { (context.variables() as MutableMap<String, String>)["new"] = "value" }
        assertThrows(UnsupportedOperationException::class.java) { (definition.parameters() as MutableMap<String, String>)["new"] = "value" }
    }

    private fun actionHandler(type: String): ScenarioActionHandler {
        return object : ScenarioActionHandler {
            override fun type(): String = type
            override fun execute(context: ScenarioExecutionContext, action: ScenarioRuntimeDefinition) {}
        }
    }
}
